File: app/src/main/java/com/ozkanmut/ilactakip/CapabilityEventGate.kt
package com.ozkanmut.ilactakip

import android.content.Context

/**
 * Deterministic ordering gate for remote capability grant/revoke events.
 * Prevents an older grant from being replayed after a newer revoke and
 * accidentally restoring access. Revision is authoritative; legacy events
 * without revisions fall back to timestamp ordering.
 */
object CapabilityEventGate {
    private const val PREFS = "dosefolk_capability_order"
    private fun prefs(c: Context) = c.getSharedPreferences(PREFS, Context.MODE_PRIVATE)

    private fun permission(event: DoseEvent): String? = when {
        event.type.startsWith("capability_edit_program_") -> "EDIT_PROGRAM"
        event.type.startsWith("capability_edit_stock_") -> "EDIT_STOCK"
        else -> null
    }

    private fun owner(event: DoseEvent): String = event.ownerId.ifBlank { event.actorTopic }
    private fun base(owner: String, permission: String) = "$owner|$permission"

    @Synchronized
    fun accept(c: Context, event: DoseEvent): Boolean {
        val permission = permission(event) ?: return false
        val owner = owner(event)
        if (owner.isBlank()) return false

        val p = prefs(c)
        val base = base(owner, permission)
        val storedRevision = p.getLong("$base|rev", 0L)
        val storedActor = p.getString("$base|actor", "").orEmpty()
        val storedEvent = p.getString("$base|event", "").orEmpty()
        val storedTimestamp = p.getLong("$base|ts", 0L)

        val newer = if (event.revision > 0L || storedRevision > 0L) {
            when {
                event.revision != storedRevision -> event.revision > storedRevision
                event.actorTopic != storedActor -> event.actorTopic > storedActor
                else -> event.eventId > storedEvent
            }
        } else {
            when {
                event.timestamp != storedTimestamp -> event.timestamp >= storedTimestamp
                event.actorTopic != storedActor -> event.actorTopic > storedActor
                else -> event.eventId > storedEvent
            }
        }
        if (!newer) return false

        p.edit()
            .putLong("$base|rev", event.revision)
            .putLong("$base|ts", event.timestamp)
            .putString("$base|actor", event.actorTopic)
            .putString("$base|event", event.eventId)
            .commit()
        return true
    }

    @Synchronized
    fun clearPeer(c: Context, topic: String) {
        if (topic.isBlank()) return
        val p = prefs(c)
        val edit = p.edit()
        p.all.keys.filter { it.startsWith("$topic|") }.forEach(edit::remove)
        edit.commit()
    }
}

File: ios/Dosefolk/Dosefolk/CapabilityEventGate.swift
import Foundation

struct CapabilityOrderingEntry: Codable, Equatable {
    var revision: Int64
    var timestamp: Int64
    var actorTopic: String
    var eventId: String
}

struct CapabilityEventGate {
    private let store: LocalStore

    init(store: LocalStore) {
        self.store = store
    }

    func accept(_ event: DoseEvent) throws -> Bool {
        guard let permission = permission(for: event) else { return false }
        let owner = event.ownerId.isEmpty ? event.actorTopic : event.ownerId
        guard !owner.isEmpty else { return false }

        let key = "\(owner)|\(permission)"
        var ordering = try store.load([String: CapabilityOrderingEntry].self, from: .capabilityOrdering, default: [:])
        if let current = ordering[key], !wins(event, over: current) {
            return false
        }

        ordering[key] = CapabilityOrderingEntry(
            revision: event.revision,
            timestamp: event.timestamp,
            actorTopic: event.actorTopic,
            eventId: event.eventId
        )
        try store.save(ordering, to: .capabilityOrdering)
        return true
    }

    func clearPeer(_ topic: String) throws {
        guard !topic.isEmpty else { return }
        var ordering = try store.load([String: CapabilityOrderingEntry].self, from: .capabilityOrdering, default: [:])
        ordering = ordering.filter { !$0.key.hasPrefix("\(topic)|") }
        try store.save(ordering, to: .capabilityOrdering)
    }

    private func permission(for event: DoseEvent) -> String? {
        if event.type.hasPrefix("capability_edit_program_") { return "EDIT_PROGRAM" }
        if event.type.hasPrefix("capability_edit_stock_") { return "EDIT_STOCK" }
        return nil
    }

    private func wins(_ event: DoseEvent, over current: CapabilityOrderingEntry) -> Bool {
        if event.revision > 0 || current.revision > 0 {
            if event.revision != current.revision { return event.revision > current.revision }
            if event.actorTopic != current.actorTopic { return event.actorTopic > current.actorTopic }
            return event.eventId > current.eventId
        }
        if event.timestamp != current.timestamp { return event.timestamp >= current.timestamp }
        if event.actorTopic != current.actorTopic { return event.actorTopic > current.actorTopic }
        return event.eventId > current.eventId
    }
}
